test/beforeHeadersSend.test.js

```diff
@@ -29,172 +32,241 @@ let setMock;
+            localStorage = {}
+            localStorage[`bypass-tokens-${config.id}`] = storedTokens
+            localStorage[`bypass-tokens-count-${config.id}`] = 2
+
+            details = {
+                method: "GET",
+                requestHeaders: [],
+                requestId: "212",
+                tabId: "101"
+            };
+            url = new URL(EXAMPLE_HREF);
+            setMockFunctions();
+            setConfig(config.id); // set the CF config
+            resetVars();
+            resetSpendVars();
+            config["spend-action"]["urls"] = [LISTENER_URLS];
+        });
+        /**
+         * Tests
+         */
+        describe(`redemptions are not attempted, config.id = ${config.id}`, () => {
+            test("redemption is off", () => {
+                workflow.__set__("DO_REDEEM", false);
+                let redeemHdrs = beforeSendHeaders(details, url);
+                expect(redeemHdrs.cancel).toBeFalsy();
+                expect(redeemHdrs.requestHeaders).toBeFalsy();
+            });
+            test(`spend flag not set, config.id = ${config.id}`, () => {
+                let redeemHdrs = beforeSendHeaders(details, url);
+                expect(redeemHdrs.cancel).toBeFalsy();
+                expect(redeemHdrs.requestHeaders).toBeFalsy();
+            });
+            test(`url is error page, config.id = ${config.id}`, () => {
+                let newUrl = EXAMPLE_HREF + "/cdn-cgi/styles/";
+                url = new URL(newUrl);
+                setSpendFlag(url.host, true);
+                let redeemHdrs = beforeSendHeaders(details, url);
+                expect(redeemHdrs.cancel).toBeFalsy();
+                expect(redeemHdrs.requestHeaders).toBeFalsy();
+            });
+            test(`url is favicon, config.id = ${config.id}`, () => {
+                let newUrl = EXAMPLE_HREF + "/favicon.ico";
+                url = new URL(newUrl);
+                setSpendFlag(url.host, true);
+                let redeemHdrs = beforeSendHeaders(details, url);
+                expect(redeemHdrs.cancel).toBeFalsy();
+                expect(redeemHdrs.requestHeaders).toBeFalsy();
+            });
+            test(`max spend has been reached, config.id = ${config.id}`, () => {
+                setSpendFlag(url.host, true);
+                setSpentHosts(url.host, 31);
+                switch (config.id) {
```

Comment:
Is this switch required? I would have thought that `SPEND_MAX` would be reset when `setConfig` was called in the `beforeEach` block above? So you should be able to just run `workflow.__get__("SPEND_MAX");`

Reply:
I'm still trying to grok how exactly `rewire` works, simple `__get__` does not seem to work. That being said, I've started a `lets-get-functional` branch that changes almost  all of the globals to function calls, it feels nicer that way. I also refactored tests there to use `jest-each`, which is much nicer than what I'm doing here. Anyway after we get this merged. I'll submit that PR, I expect to do a lot of work around tests in general there, extract globals and cleanup in general.